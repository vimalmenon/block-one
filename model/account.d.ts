export interface AccountParams {
  customerId: string
  accountNo: string
}

export interface AccountModel {
  customerId: number
  accountNumber: string
  accountType: string
  balance: number
}
