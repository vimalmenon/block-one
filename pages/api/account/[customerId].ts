import type { NextApiRequest, NextApiResponse } from 'next';
import {AccountModel} from "../../../model/account";

import {openDb} from "../../../database.setup";

export default async (req: NextApiRequest, res: NextApiResponse<AccountModel[]>) => {
    if (req.method ==="GET") {
        const {customerId, transfer} = req.query;
        let result;
        if (transfer) {
            result = await (await openDb()).all('SELECT customer_id as customerId, account_number as accountNumber, account_type as accountType, balance FROM accounts WHERE customer_id = ? and account_type IN ("Saving Account", "Checking Account")', customerId);
        } else {
            result = await (await openDb()).all('SELECT customer_id as customerId, account_number as accountNumber, account_type as accountType, balance FROM accounts WHERE customer_id = ?', customerId);
        }

        res.status(200).json(result);
    }
}