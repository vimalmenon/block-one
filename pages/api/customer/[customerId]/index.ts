import type { NextApiRequest, NextApiResponse } from 'next'

import { openDb } from '../../../../database.setup'

export default async (req: NextApiRequest, res: NextApiResponse<any>) => {
  if (req.method === 'GET') {
    const { customerId } = req.query
    const result = await (
      await openDb()
    ).all(
      'SELECT prefix, name, customer_id as customerId, address as address, email_address as emailAddress,  contact_no as contactNo FROM customers WHERE customer_id = ? LIMIT 1',
      customerId
    )
    res.status(200).json(result[0])
  }
}
