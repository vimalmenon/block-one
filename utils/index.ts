export * from "./navigate";
export * from "./convertCurrency";