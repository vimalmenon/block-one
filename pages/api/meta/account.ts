import type { NextApiRequest, NextApiResponse } from 'next';
import {AccountParams} from "../../../model/account";

import {openDb} from "../../../database.setup";

export default async (req: NextApiRequest, res: NextApiResponse<AccountParams[]>) => {
  if (req.method==="GET") {
    const results = await (await openDb()).all("select customer_id as customerId,account_number as accountNo  from accounts");
    const params:AccountParams[] = []
    results.forEach((result) => {
      params.push({
        customerId: result.customerId.toString(),
        accountNo: result.accountNo
      });
    });
    res.status(200).json(params)
  }
}