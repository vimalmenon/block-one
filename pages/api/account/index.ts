import type { NextApiRequest, NextApiResponse } from 'next';
import {AccountModel} from "../../../model/account";
import {CustomerModel} from "../../../model/customer"

//Last Login


const customerData:CustomerModel[] = [
    {
        prefix:"Mrs",
        name:"Jane Doe",
        customerId: 12345,
        address:"Kowloon, HK",
        emailAddress:"jane.doe@example.com",
        contactNo:"1234 7889"
    },
    {
        prefix:"Mr",
        name:"John Doe",
        customerId: 67890,
        address:"Central, HK",
        emailAddress:"John.doe@example.com",
        contactNo:"7889 1234"
    }
];


const customerAccountDetails:AccountModel[] = [
    {
        customerId:12345,
        accountNumber: "123-456-789",
        accountType: "Saving Account",
        balance:20000
    },
    {
        customerId:12345,
        accountNumber: "123-456-788",
        accountType: "Checking Account",
        balance:10000
    },
    {
        customerId:12345,
        accountNumber: "123-456-788",
        accountType: "Current Account",
        balance:5000
    },
    {
        customerId:12345,
        accountNumber: "123-456-787",
        accountType: "MPF Account",
        balance:2000
    },
    {
        customerId:67890,
        accountNumber: "123-456-786",
        accountType: "Saving Account",
        balance:40000
    },
    {
        customerId:67890,
        accountNumber: "123-456-785",
        accountType: "Checking Account",
        balance:30000
    },
    {
        customerId:67890,
        accountNumber: "123-456-788",
        accountType: "Current Account",
        balance:8000
    },
    {
        customerId:67890,
        accountNumber: "123-456-784",
        accountType: "MPF Account",
        balance:3000
    }
]

export default (req: NextApiRequest, res: NextApiResponse<any>) => {
    if (req.method ==="GET") {
        const {accountId} = req.query;
        const selectedCustomer = customerAccountDetails.filter((customer) => {
            return customer.customerId.toString() === accountId;
        })
        res.status(200).json(selectedCustomer);
    }
  
}