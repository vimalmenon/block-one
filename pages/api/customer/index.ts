import type { NextApiRequest, NextApiResponse } from 'next'
import { CustomerModel } from '../../../model/customer'

import { openDb } from '../../../database.setup'

export default async (req: NextApiRequest, res: NextApiResponse<CustomerModel[]>) => {
  if (req.method === 'GET') {
    const results = await (
      await openDb()
    ).all(
      'select prefix, name, customer_id as customerId, address, email_address as emailAddress, contact_no as contactNo from customers'
    )
    res.status(200).json(results)
  }
}
