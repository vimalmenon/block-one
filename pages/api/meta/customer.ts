import type { NextApiRequest, NextApiResponse } from 'next'
import { CustomerParams } from '../../../model/customer'

import { openDb } from '../../../database.setup'

export default async (req: NextApiRequest, res: NextApiResponse<CustomerParams[]>) => {
  if (req.method === 'GET') {
    const results = await (await openDb()).all('select customer_id as customerId FROM accounts')
    const params: CustomerParams[] = []
    results.forEach((result) => {
      params.push({
        customerId: result.customerId.toString(),
      })
    })
    res.status(200).json(params)
  }
}
