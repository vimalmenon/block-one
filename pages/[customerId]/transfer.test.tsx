import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { AccountModel } from '../../model/account'
import { CustomerModel } from '../../model/customer'
import { onLinkNavigation } from '../../utils'
import React from 'react'

import Transfer from './transfer'

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: {
      customerId: '12345',
    },
  }),
}))

jest.mock('../../utils', () => ({
  onLinkNavigation: jest.fn(),
}))

describe('Transfer Page', () => {
  test('Transfer: Snapshot testing', () => {
    const customer: CustomerModel = {
      prefix: 'Mr',
      name: 'test value',
      customerId: 12345,
      address: 'test Address',
      emailAddress: 'test.value@test.com',
      contactNo: '123-456-789',
    }

    const accounts: AccountModel[] = [
      {
        customerId: 1234,
        accountNumber: '12233344',
        accountType: 'Saving Account',
        balance: 20000,
      },
    ]
    const wrapper = shallow(<Transfer customer={customer} accounts={accounts} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  test('Transfer: Snapshot testing with fromAccount', () => {
    const customer: CustomerModel = {
      prefix: 'Mr',
      name: 'test value',
      customerId: 12345,
      address: 'test Address',
      emailAddress: 'test.value@test.com',
      contactNo: '123-456-789',
    }

    const accounts: AccountModel[] = [
      {
        customerId: 1234,
        accountNumber: '12233344',
        accountType: 'Saving Account',
        balance: 20000,
      },
    ]
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [accounts, jest.fn()])
      .mockImplementationOnce(() => [accounts[0], jest.fn()])
    const wrapper = shallow(<Transfer customer={customer} accounts={accounts} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  test('Transfer: Testing of Back button', () => {
    const customer: CustomerModel = {
      prefix: 'Mr',
      name: 'test value',
      customerId: 12345,
      address: 'test Address',
      emailAddress: 'test.value@test.com',
      contactNo: '123-456-789',
    }

    const accounts: AccountModel[] = [
      {
        customerId: 12345,
        accountNumber: '12233344',
        accountType: 'Saving Account',
        balance: 20000,
      },
    ]
    const wrapper = shallow(<Transfer customer={customer} accounts={accounts} />)
    wrapper.find('[data-testid="back"]').simulate('click')
    expect(onLinkNavigation).toHaveBeenCalled()
    expect(onLinkNavigation).toHaveBeenCalledWith(`/12345`)
  })
  test('Transfer: Testing of Transfer button', () => {
    const customer: CustomerModel = {
      prefix: 'Mr',
      name: 'test value',
      customerId: 12345,
      address: 'test Address',
      emailAddress: 'test.value@test.com',
      contactNo: '123-456-789',
    }

    const accounts: AccountModel[] = [
      {
        customerId: 12345,
        accountNumber: '12233344',
        accountType: 'Saving Account',
        balance: 20000,
      },
    ]
    const wrapper = shallow(<Transfer customer={customer} accounts={accounts} />)
    wrapper.find('[data-testid="transfer"]').simulate('click')
    expect(onLinkNavigation).toHaveBeenCalled()
    expect(onLinkNavigation).toHaveBeenCalledWith(`/12345`)
  })
})
