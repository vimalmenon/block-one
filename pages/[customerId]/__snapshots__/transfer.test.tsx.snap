// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transfer Page Transfer: Snapshot testing 1`] = `
<Container
  customer={
    Object {
      "address": "test Address",
      "contactNo": "123-456-789",
      "customerId": 12345,
      "emailAddress": "test.value@test.com",
      "name": "test value",
      "prefix": "Mr",
    }
  }
>
  <MetaData
    title="Transfer money between account for test value"
  />
  <div
    className="my-6 mx-4 flex w-full flex-col py-10"
  >
    <div
      className="flex justify-center"
    >
      <AccountContainer
        accounts={
          Array [
            Object {
              "accountNumber": "12233344",
              "accountType": "Saving Account",
              "balance": 20000,
              "customerId": 1234,
            },
          ]
        }
        isFromAccount={true}
        onAccountSelect={[Function]}
        setTransferAmount={[Function]}
        title="From Account"
      />
      <div
        className="flex w-6/12 flex-col"
      />
    </div>
    <div
      className="flex my-4 justify-between"
    >
      <Button
        data-testid="transfer"
        disabled={true}
        onClick={[Function]}
      >
        Transfer
      </Button>
      <Button
        data-testid="back"
        onClick={[Function]}
        type="secondary"
      >
        Back
      </Button>
    </div>
  </div>
</Container>
`;

exports[`Transfer Page Transfer: Snapshot testing with fromAccount 1`] = `
<Container
  customer={
    Object {
      "address": "test Address",
      "contactNo": "123-456-789",
      "customerId": 12345,
      "emailAddress": "test.value@test.com",
      "name": "test value",
      "prefix": "Mr",
    }
  }
>
  <MetaData
    title="Transfer money between account for test value"
  />
  <div
    className="my-6 mx-4 flex w-full flex-col py-10"
  >
    <div
      className="flex justify-center"
    >
      <AccountContainer
        accounts={
          Array [
            Object {
              "accountNumber": "12233344",
              "accountType": "Saving Account",
              "balance": 20000,
              "customerId": 1234,
            },
          ]
        }
        isFromAccount={true}
        onAccountSelect={[Function]}
        selectedAccount={
          Object {
            "accountNumber": "12233344",
            "accountType": "Saving Account",
            "balance": 20000,
            "customerId": 1234,
          }
        }
        setTransferAmount={[Function]}
        title="From Account"
      />
      <div
        className="flex w-1/12 flex-col items-center justify-center"
      >
        <svg
          className="h-6 w-6 text-blue-500"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 7h12m0 0l-4-4m4 4l-4 4m0 6H4m0 0l4 4m-4-4l4-4"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
          />
        </svg>
      </div>
      <AccountContainer
        accounts={
          Array [
            Object {
              "accountNumber": "12233344",
              "accountType": "Saving Account",
              "balance": 20000,
              "customerId": 1234,
            },
          ]
        }
        onAccountSelect={[Function]}
        selectedAccount={
          Object {
            "accountNumber": "12233344",
            "accountType": "Saving Account",
            "balance": 20000,
            "customerId": 1234,
          }
        }
        title="To Account"
      />
    </div>
    <div
      className="flex my-4 justify-between"
    >
      <Button
        data-testid="transfer"
        disabled={true}
        onClick={[Function]}
      >
        Transfer
      </Button>
      <Button
        data-testid="back"
        onClick={[Function]}
        type="secondary"
      >
        Back
      </Button>
    </div>
  </div>
</Container>
`;
